From c5ada331bec79142c85a5cddb49d538a2c3c8651 Mon Sep 17 00:00:00 2001
From: Marcell Brauner <bmarcell@inf.u-szeged.hu>
Date: Tue, 9 Sep 2025 14:15:13 +0200
Subject: [PATCH 12/13] UPSTREAM: Disable GBM if Ozone X11 is not available

If certain X11 libraries are missing then the BUILDFLAG(IS_OZONE_X11) is
false. In this case, the GLX code path is unable to create GBM buffer
without Ozone X11 support. This results in the web view being empty and
the warning messages are not helpful.

To circumvent this, add a more meaningful warning message and fallback
to a different rendering path.

Pick-to: 6.10 6.9 6.8
Change-Id: Id4bc9a7e88c6f19713c764437138b92a3aee86e6
Reviewed-by: Moss Heim <moss.heim@qt.io>
Reviewed-by: Allan Sandfeld Jensen <allan.jensen@qt.io>
---
 qtwebengine/src/core/web_engine_context.cpp | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/qtwebengine/src/core/web_engine_context.cpp b/qtwebengine/src/core/web_engine_context.cpp
index e5d3cd8385..5905b6c379 100644
--- a/qtwebengine/src/core/web_engine_context.cpp
+++ b/qtwebengine/src/core/web_engine_context.cpp
@@ -1207,12 +1207,20 @@ bool WebEngineContext::isGbmSupported()
             return false;
         }
 
+#if !BUILDFLAG(IS_OZONE_X11)
+        if (OzoneUtilQt::usingGLX()) {
+            qWarning("GLX: Disable GBM because Ozone X11 is not available. "
+                     "Possibly caused by missing libraries for qpa-xcb support.");
+            return false;
+        }
+#endif
+
         return true;
     }();
 
     return supported;
 }
-#endif
+#endif // BUILDFLAG(IS_OZONE)
 
 void WebEngineContext::registerMainThreadFactories()
 {
-- 
2.51.0

